actors/hierarchical_sca/src/atomic/mod.rs

```diff
@@ -0,0 +1,150 @@
+use cid::multihash::Code::Blake2b256;
+use cid::multihash::MultihashDigest;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use fvm_ipld_encoding::{serde_bytes, tuple::*, Cbor, RawBytes, DAG_CBOR};
+use fvm_shared::MethodNum;
+use serde::de::DeserializeOwned;
+use serde::ser::Serialize;
+
+use crate::tcid::{TCid, THamt};
+
+/// MethodNum to lock some state in an actor
+/// This methods are only supported in actors
+/// that support atomic executions.
+pub const METHOD_LOCK: MethodNum = 2;
+/// MethodNum used to trigger the merge of an input with
+/// other input locked states.
+pub const METHOD_MERGE: MethodNum = 3;
+/// MethodNum called to signal the abortion of an atomic execution
+/// and the unlock of all locked states in the actor for the execution
+pub const METHOD_ABORT: MethodNum = 4;
+/// MethodNum to trigger the merge of the output of an execution
+/// into the state of an actor, and the unlock of all locked states.
+pub const METHOD_UNLOCK: MethodNum = 5;
+
+/// Trait that determines the functions that need to be implemented by
+/// a state object to be lockable and be used in an atomic execution.
+///
+/// Different strategies may be used to merge different locked state to
+/// prepare the actor state for the execution, and for the merging of the
+/// output of the execution to the original state of the actor.
+pub trait LockableState<S: Serialize + DeserializeOwned> {
+    /// Merge a locked state (not necessarily the output) to the current state.
+    fn merge(other: Self) -> anyhow::Result<()>;
+    /// Merge the output of an execution to the current state.
+    fn merge_output(other: Self) -> anyhow::Result<()>;
```

Comment:
```suggestion
    /// Merge a locked state (not necessarily the output) to the current state.
    fn merge(&mut self, other: Self) -> anyhow::Result<()>;
    /// Merge the output of an execution to the current state.
    fn merge_output(&mut self, other: Self) -> anyhow::Result<()>;
```
It's definitely missing an argument, and judging by the return value it wants to mutate it.

There are other alternatives that would be more functional:
```
    /// Merge a locked state (not necessarily the output) to the current state.
    fn merge(self, other: Self) -> anyhow::Result<Self>;
    /// Merge the output of an execution to the current state.
    fn merge_output(self, other: Self) -> anyhow::Result<Self>;
```

Another consideration would be whether to return the `other` as part of the error, or always consider it consumed even if it failed to merge.

Reply:
Good call! Thank you for also reviewing this file, actually these were just placeholders for the next PR where I'll add all the low-level primitives for the atomic execution, but I will fix it right away. For now I'll go with the first option (internal mutation), but it'll probably change in the next PR (once we have also a sample actor to play with and see how usable the trait is).